File: 1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
  public int reverseInOrderTraversal(TreeNode root, int sum)
  {
    if(root == null)
      return sum;
    int right = reverseInOrderTraversal(root.right, sum);
    root.val += right;
    int left = reverseInOrderTraversal(root.left, root.val);
    return left;
  }
  
  public TreeNode bstToGst(TreeNode root) {
    reverseInOrderTraversal(root, 0);
    return root;
  }
}

File: 1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

void reverseInOrderTraversal(TreeNode* root, int& sum){
  if(!root)
    return;
  reverseInOrderTraversal(root->right, sum);
  sum+= root->val;
  root->val = sum;
  reverseInOrderTraversal(root->left, sum);
}

class Solution {
public:
  TreeNode* bstToGst(TreeNode* root) {
    int sum=0;
    reverseInOrderTraversal(root, sum);
    return root;
  }
};